server/src/main/java/org/elasticsearch/indices/recovery/plan/ShardSnapshotsService.java

```diff
@@ -0,0 +1,124 @@
+import org.elasticsearch.action.admin.cluster.snapshots.get.shard.GetShardSnapshotAction;
+import org.elasticsearch.action.admin.cluster.snapshots.get.shard.GetShardSnapshotRequest;
+import org.elasticsearch.action.admin.cluster.snapshots.get.shard.GetShardSnapshotResponse;
+import org.elasticsearch.action.support.ThreadedActionListener;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.metadata.RepositoriesMetadata;
+import org.elasticsearch.cluster.metadata.RepositoryMetadata;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.blobstore.BlobContainer;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot;
+import org.elasticsearch.indices.recovery.RecoverySettings;
+import org.elasticsearch.repositories.RepositoriesService;
+import org.elasticsearch.repositories.ShardSnapshotInfo;
+import org.elasticsearch.repositories.blobstore.BlobStoreRepository;
+import org.elasticsearch.snapshots.Snapshot;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.indices.recovery.RecoverySettings.SNAPSHOT_RECOVERIES_SUPPORTED_VERSION;
+
+public class ShardSnapshotsService {
+    private final Logger logger = LogManager.getLogger(ShardSnapshotsService.class);
+
+    private final Client client;
+    private final RepositoriesService repositoriesService;
+    private final ThreadPool threadPool;
+    private final ClusterService clusterService;
+
+    public ShardSnapshotsService(Client client,
+                                 RepositoriesService repositoriesService,
+                                 ThreadPool threadPool,
+                                 ClusterService clusterService) {
+        this.client = client;
+        this.repositoriesService = repositoriesService;
+        this.threadPool = threadPool;
+        this.clusterService = clusterService;
+    }
+
+    public void fetchLatestSnapshotsForShard(ShardId shardId, ActionListener<List<ShardSnapshot>> listener) {
+        if (shardId == null) {
+            throw new IllegalArgumentException("SharId was null but a value was expected");
+        }
+
+        final RepositoriesMetadata currentReposMetadata = clusterService.state()
+            .metadata()
+            .custom(RepositoriesMetadata.TYPE, RepositoriesMetadata.EMPTY);
+
+        List<String> repositories = currentReposMetadata.repositories()
+            .stream()
+            .filter(repositoryMetadata -> RecoverySettings.REPOSITORY_SNAPSHOT_BASED_RECOVERY_SETTING.get(repositoryMetadata.settings()))
+            .map(RepositoryMetadata::name)
+            .collect(Collectors.toList());
+
+        if (repositories.isEmpty() || masterSupportsFetchingLatestSnapshots() == false) {
```

Comment:
I reckon we might find it useful to add some `DEBUG` logging indicating which path through this method we took (i.e. to distinguish no good repositories vs cluster too old vs actually trying to search repositories).